src/lib/x509/x509_ext.cpp

```diff
@@ -868,6 +877,582 @@ const std::string& TNAuthList::Entry::telephone_number() const {
+   return fam;
+}
+
+void IPAddressBlocks::IPAddressFamily::encode_into(Botan::DER_Encoder& into) const {
+   into.start_sequence();
+
+   std::vector<uint8_t> afam = {static_cast<uint8_t>(m_afi >> 8), static_cast<uint8_t>(m_afi & 0xff)};
+
+   if(m_safi.has_value()) {
+      afam.push_back(m_safi.value());
+   }
+
+   into.add_object(ASN1_Type::OctetString, ASN1_Class::Universal, afam);
+
+   if(std::holds_alternative<IPAddressChoice<Version::IPv4>>(m_ip_addr_choice)) {
+      into.encode(std::get<IPAddressChoice<Version::IPv4>>(m_ip_addr_choice));
+   } else {
+      into.encode(std::get<IPAddressChoice<Version::IPv6>>(m_ip_addr_choice));
+   }
+   into.end_cons();
+}
+
+void IPAddressBlocks::IPAddressFamily::decode_from(Botan::BER_Decoder& from) {
+   BER_Object obj = from.get_next_object();
+   ASN1_Type type_tag = obj.type_tag();
+   if(type_tag != ASN1_Type::Sequence) {
+      throw Decoding_Error(fmt("Unexpected type for IPAddressFamily {}", static_cast<uint32_t>(type_tag)));
+   }
+   BER_Decoder obj_ber = BER_Decoder(obj);
+
+   std::vector<uint8_t> addr_family;
+   obj_ber.decode(addr_family, ASN1_Type::OctetString);
+   size_t addr_family_length = addr_family.size();
+
+   if(addr_family_length != 2 && addr_family_length != 3) {
+      throw Decoding_Error("(S)AFI can only contain 2 or 3 bytes");
+   }
+
+   m_afi = (addr_family[0] << 8) | addr_family[1];
+
+   if(addr_family_length == 3) {
+      m_safi = addr_family[2];
+   }
+
+   if(m_afi == 1) {
+      IPAddressChoice<Version::IPv4> addr_choice;
+      obj_ber.decode(addr_choice);
+      m_ip_addr_choice = addr_choice;
+   } else if(m_afi == 2) {
+      IPAddressChoice<Version::IPv6> addr_choice;
+      obj_ber.decode(addr_choice);
+      m_ip_addr_choice = addr_choice;
+   } else {
+      throw Decoding_Error("Only AFI IPv4 and IPv6 are supported.");
+   }
+}
+
+template <IPAddressBlocks::Version V>
+void IPAddressBlocks::IPAddressChoice<V>::encode_into(Botan::DER_Encoder& into) const {
+   if(m_ip_addr_ranges.has_value()) {
+      into.start_sequence().encode_list(m_ip_addr_ranges.value()).end_cons();
+   } else {
+      into.encode_null();
+   }
+}
+
+template <IPAddressBlocks::Version V>
+void IPAddressBlocks::IPAddressChoice<V>::decode_from(Botan::BER_Decoder& from) {
+   ASN1_Type next_tag = from.peek_next_object().type_tag();
+
+   if(next_tag == ASN1_Type::Null) {
+      from.decode_null();
+      m_ip_addr_ranges = std::nullopt;
+   } else if(next_tag == ASN1_Type::Sequence) {
+      std::vector<IPAddressOrRange<V>> ip_ranges;
+      from.decode_list(ip_ranges);
+      m_ip_addr_ranges = ip_ranges;
+   } else {
+      throw Decoding_Error(fmt("Unexpected type for IPAddressChoice {}", static_cast<uint32_t>(next_tag)));
+   }
+}
+
+template <IPAddressBlocks::Version V>
+void IPAddressBlocks::IPAddressChoice<V>::sort_and_merge() {
+   if(!m_ip_addr_ranges.has_value()) {
+      return;
+   }
+   std::vector<IPAddressOrRange<V>> current = m_ip_addr_ranges.value();
```

Comment:
This could be captured by reference

`const auto& current = m_ip_addr_ranges.value()`